presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,271 @@
+            cast = (Cast) expression.getLeft();
+            literal = (Literal) expression.getRight();
+            operator = expression.getOperator();
+        }
+        else if (expression.getLeft() instanceof Literal && expression.getRight() instanceof Cast) {
+            cast = (Cast) expression.getRight();
+            literal = (Literal) expression.getLeft();
+            operator = expression.getOperator().flip();
+        }
+        else {
+            return expression;
+        }
+
+        if (literal instanceof NullLiteral) {
+            return expression;
+        }
+
+        Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+        Type literalType = typeAnalyzer.getType(session, types, literal);
+
+        Optional<Type.Range> range = sourceType.getRange();
+        if (!range.isPresent()) {
+            return expression;
+        }
+
+        Object min = range.get().getMin();
+        Object max = range.get().getMax();
+
+        Object maxInTargetType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
+                session.toConnectorSession(),
+                max);
+
+        Object minInTargetType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
+                session.toConnectorSession(),
+                min);
+
+        Object literalValue = ExpressionInterpreter.evaluateConstantExpression(literal, literalType, metadata, session, ImmutableList.of());
+
+        int upperBoundComparison = compare(literalType, literalValue, maxInTargetType);
+
+        if (upperBoundComparison > 0) {
+            // larger than maximum representable value
+            switch (operator) {
+                case EQUAL:
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                    return falseIfNotNull(cast.getExpression());
+                case NOT_EQUAL:
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    return trueIfNotNull(cast.getExpression());
+                case IS_DISTINCT_FROM:
+                    return TRUE_LITERAL;
+            }
+        }
+
+        if (upperBoundComparison == 0) {
+            // equal to max representable value
+            switch (operator) {
+                case GREATER_THAN:
+                    return falseIfNotNull(cast.getExpression());
+                case GREATER_THAN_OR_EQUAL:
+                    return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(max, sourceType));
+                case LESS_THAN_OR_EQUAL:
+                    return trueIfNotNull(cast.getExpression());
+                case LESS_THAN:
+                case EQUAL:
+                case IS_DISTINCT_FROM:
+                    return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(max, sourceType));
```

Comment:
Same here. Consider `x` of `decimal(1,1)` type and `exp`: `NOT( cast(x as integer) IS DISTINCT FROM 1 )`.
Both `0.7` and `0.9` satisfy `exp`.